### ISSUE TITLE
Serialization of case class with overridden attributes not working 

### ISSUE THREAD

**Issue by karaboM** [FasterXML/jackson-module-scala#218]
When serializing a case class with overridden attributes, the attributes in question are not being serialized. So given the following code

``` scala
abstract class A(val aa: Seq[X]){
  def isGG: Boolean = false
}

case class B(override val aa: Seq[X]) extends A(aa){
  override def isGG: Boolean = true
}

abstract class X(val a: String, val b: String, val c: String, val d: Option[Map[String, String]])

case class Y(override val a: String, override val b: String, override val c: String,
              override val d: Option[Map[String, String]]) extends X(a, b, c, d)
```

 which is the definition of the case class to be serialized class B, and the overridden attributed that will not be serialized aa. When executing the serialization code below:

``` scala
import com.fasterxml.jackson.databind.{DeserializationFeature, SerializationFeature, ObjectMapper}
import com.fasterxml.jackson.datatype.joda.JodaModule
import com.fasterxml.jackson.module.scala.DefaultScalaModule

object Program extends App{

  val y = Y("1","2","3",None)
  val b = B(Seq(y))

  val json = serializeJson(b)
  json

  def serializeJson(b: B) = {
    val objectMapper = new ObjectMapper()
    objectMapper.registerModule(new DefaultScalaModule)
    objectMapper.writeValueAsString(b)
  }
}
```

the output of this results in the following value for the json attribute:
"{
  "gg" : true
}
"

which should be 
"
{
 "aa": ..................
  "gg" : true
}
"

This is present in version 2.5.3 and 2.6.0 version

**Comment by christophercurrie** [comment]
Interesting. I didn't know that an overridden val would change its default visibility. I'm not sure which part of the Jackson visibility checking might be failing here, or what the JVM implementation looks like (as I though the JVM wouldn't allow visibility changes like this for methods). I'll dig into it and see what I can learn.

**Comment by karaboM** [comment]
Thanks Chris!!!

**Comment by shannoncolavita** [comment]
Hi Chris is this issue going to be looked at for the next release? 
Thanks

**Comment by Jinxbel** [comment]
We've just run into this issue.
it maybe part of the cause but if you look at how reflection looks at this, something seems a bit weird.

In the original example B.aa is reflected as a method (found in getDeclaredMethods) and A.aa is declared as a Field (found in getDeclaredFields).

Could this be the cause?

**Comment by kdubezerra** [comment]
Just ran into the exact same issue... Any plans to fix this?

```
trait MyTrait {
    val id: UUID
    val `type`: String
}

class Base(val id: UUID) extends MyTrait {
    override val `type`: String = "baseclass"
}

case class Sub(override val id: UUID, something: String) extends Base(id)
```

Is there some workaround to prevent this issue from manifesting at least?

**Comment by shannoncolavita** [comment]
The workaround is to omit the override keyword
sealed trait MyTrait {
  val id: UUID
  val `type`: String
}

trait Base extends MyTrait {
  val `type`: String = "baseclass"
}

case class Sub(id: UUID, something: String) extends Base

**Comment by shannoncolavita** [comment]
The workaround is to omit the override keyword
`sealed trait MyTrait {
  val id: UUID
  val `type`: String
}

trait Base extends MyTrait {
  val `type`: String = "baseclass"
}

case class Sub(id: UUID, something: String) extends Base`

**Comment by kdubezerra** [comment]
Thanks for the reply, Shannon. The problem with this approach is that I cannot instantiate Base anymore (unless I create a fourth class that implements it)... My "solution" was not to extend Base, and repeat the `type` attribute in Sub... Perhaps your solution, plus a fourth class, would be less ugly...

**Comment by shannoncolavita** [comment]
No problem... We basically redesigned our classes to get around this.  I guess another option is to submit a PR :)

**Comment by talbenshabtay** [comment]
came across this now , still happening for version - 2.8.7
any result other than Refactoring code ? 

btw my solution is just to drop the override val keywords and pass it along the parent CTOR.

**Comment by msilvestre** [comment]
So more than 3 years passed, and this issue is still present. Just come across it now.. :(

With Jackson 2.10.0

**Comment by kupci** [comment]
Do the workarounds in this thread solve the issue? Also note Shannon's [comment](https://github.com/FasterXML/jackson-module-scala/issues/218#issuecomment-258562288):

> I guess another option is to submit a PR :)

**Comment by cowtowncoder** [comment]
@msilvestre yes, despite your best in helping resolve this issue, it unfortunately still exists. Can not think why this would be. Typically regular follow-up on git issue will over time invariably result in fixes.

**Comment by pjfanning** [comment]
I have made a change to 2.10 code which should be released with 2.10.3. The unit test for code change is based on https://github.com/FasterXML/jackson-module-scala/issues/218#issuecomment-258498870. The change doesn't break existing unit tests and hopefully shouldn't have side effects for other use cases.

**Comment by pjfanning** [comment]
this fix appears to cause https://github.com/swagger-akka-http/swagger-akka-http/issues/171
